### PR TITLE
DrivAerML: WD=1e-2 + gc=0.5 compound regularization at 4L/512d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

On AirfRANS, the winning recipe uses WD=1e-2 + gc (0.5 or 1.0) — both regularization mechanisms working together. The DrivAerML baseline uses WD=1e-4 (100x lower) and no gradient clipping. Your previous experiment (#2793) showed gc=1.5 is too loose for DrivAerML's low natural grad norms (~0.76-0.85). This experiment tests the AirfRANS-proven compound: WD=1e-2 + gc=0.5 — strong regularization + tight clipping matched to the gradient scale. We also test WD=1e-2 alone and gc=0.5 alone to measure the individual vs compound effect.

Note: taki (#2814) has WD=1e-3 + gc=0.5 (10x less WD). fern (#2794) has WD=1e-2 alone (no gc). This fills the matrix gap: WD=1e-2 + gc=0.5.

## Instructions

Run 4 experiments in parallel (2 GPUs each):

**Run 1 — WD=1e-2 + gc=0.5 (main compound)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0,1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-wd-gc \
  --wandb-name driv-wd1e2-gc05
```

**Run 2 — WD=1e-2 + gc=1.0 (looser clip)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2,3 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-wd-gc \
  --wandb-name driv-wd1e2-gc10
```

**Run 3 — WD=5e-3 + gc=0.5 (intermediate WD)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4,5 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --weight-decay 5e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-wd-gc \
  --wandb-name driv-wd5e3-gc05
```

**Run 4 — WD=1e-2 + gc=0.5 + lr=7e-4 (higher LR compound)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=6,7 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw \
  --lr 7e-4 --cosine-t-max 30 --grad-clip 0.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-wd-gc \
  --wandb-name driv-wd1e2-gc05-lr7e4
```

Report best val_primary/surface_rel_l2_pct checkpoint for each run. Compare against baseline 4.619% and note whether the compound outperforms the individual components (fern's WD-only and ymir/kaneda's gc-only runs).

## Baseline

- **Primary metric:** val_primary/surface_rel_l2_pct (lower is better)
- **Current best:** 4.619% (PR #2691, 4L/512d, AdamW lr=5e-4, T_max=30, WD=1e-4, no gc)
- **External target:** 3.71% (1.24x gap remaining)
- **AirfRANS analogy:** WD=1e-2 + gc=1.0 at 3L/256d achieved val=0.001479 (65.6% below external target)
- **Your insight from #2793:** DrivAerML grad norms ~0.76-0.85, so gc=0.5 will clip ~every step, providing continuous regularization unlike gc=1.5

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary

All 4 compound regularization runs are **significantly worse** than the baseline 4.619%. Strong WD + gc hurts DrivAerML — the AirfRANS recipe does not transfer.

| Run | Config | Best Val (%) | Epoch | Test (%) | W&B Run ID |
|-----|--------|-------------|-------|----------|------------|
| 1 | WD=1e-2 + gc=0.5 | **8.264** | 84/95 | 8.819 | 1ka0b5gp |
| 2 | WD=1e-2 + gc=1.0 | **8.093** | 91/95 | 9.458 | a1rw083y |
| 3 | WD=5e-3 + gc=0.5 | **7.925** | 95/96 | 9.571 | p7ns51bd |
| 4 | WD=1e-2 + gc=0.5 + lr=7e-4 | **9.055** | 91/93 | 10.529 | 6wrzgna8 |

**Baseline:** 4.619% (PR #2691, WD=1e-4, no gc)

### Key Metrics

- **Best run:** Run 3 (WD=5e-3 + gc=0.5) at 7.925% — still 1.72x worse than baseline
- **All runs:** 7.9-9.1% vs 4.619% baseline — compound regularization roughly doubles the error
- **Epochs:** 93-96 epochs in 180 min (~1.9 min/epoch), comparable to baseline's 267 epochs (~0.67 min/ep) — runs were significantly slower (2.8x fewer epochs)
- **Grad norms:** 0.42-0.62 mean (vs baseline ~0.76-0.85 without gc), gc=0.5 was clipping 150-280 steps/epoch, gc=1.0 only clipping ~29 steps/epoch

### Gradient Clipping Analysis

| Run | gc | Clip events (best ep) | Mean grad norm (best ep) |
|-----|----|-----------------------|--------------------------|
| 1 | 0.5 | 211/394 (54%) | 0.621 |
| 2 | 1.0 | 29/394 (7%) | 0.541 |
| 3 | 0.5 | 166/394 (42%) | 0.544 |
| 4 | 0.5 | 120/394 (30%) | 0.422 |

gc=0.5 clips a substantial fraction of batches. gc=1.0 barely clips at all since natural grad norms are ~0.5-0.6 (lower than the 0.76-0.85 reported in #2793, likely because WD=1e-2 suppresses weight magnitude).

### Commands Used

```bash
# Run 1: WD=1e-2 + gc=0.5
CUDA_VISIBLE_DEVICES=0,1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-wd-gc --wandb-name driv-wd1e2-gc05

# Run 2: WD=1e-2 + gc=1.0
CUDA_VISIBLE_DEVICES=2,3 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-wd-gc --wandb-name driv-wd1e2-gc10

# Run 3: WD=5e-3 + gc=0.5
CUDA_VISIBLE_DEVICES=4,5 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --weight-decay 5e-3 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-wd-gc --wandb-name driv-wd5e3-gc05

# Run 4: WD=1e-2 + gc=0.5 + lr=7e-4
CUDA_VISIBLE_DEVICES=6,7 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-wd-gc --wandb-name driv-wd1e2-gc05-lr7e4
```

### What Happened

**The AirfRANS compound regularization recipe catastrophically fails on DrivAerML.** All 4 runs are ~1.7-2x worse than the baseline.

Two mechanisms explain the failure:

1. **Throughput penalty:** The runs completed only 93-96 epochs in 180 min (~1.9 min/epoch) vs the baseline's 267 epochs (~0.67 min/ep). The 2-GPU allocation (to run 4 experiments in parallel) plus regularization overhead cut throughput by ~2.8x. DrivAerML's baseline wins through sheer epoch count at 4L/512d — anything that reduces throughput is fighting the dominant training-time lever.

2. **WD=1e-2 is too aggressive for DrivAerML:** The baseline uses WD=1e-4. Jumping to 1e-2 (100x) suppresses weight magnitude too aggressively, lowering grad norms from 0.76-0.85 to 0.42-0.62. This over-regularization prevents the model from fitting the complex 3D automotive flow field. Even WD=5e-3 (Run 3) was too strong.

The key difference from AirfRANS: AirfRANS has a much simpler geometry (2D airfoil) where over-regularization can push toward good generalization, while DrivAerML's complex 3D geometry needs more model capacity/flexibility.

### Suggested Follow-ups

- **WD=1e-3 + gc=0.5 on 1 GPU:** taki (#2814) is already testing WD=1e-3, so wait for those results first. If WD=1e-3 alone helps, then compound with gc=0.5 at that gentler WD level.
- **gc=0.5 alone (no WD increase):** Test gc=0.5 at the baseline WD=1e-4 to isolate the clipping effect. This would show whether gentle gradient clipping helps at all when WD is already low.
- **The throughput confound:** Re-run the best config (WD=5e-3 + gc=0.5) on a single GPU to get ~267 epochs and see if longer training partially closes the gap.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2832 because ADVISOR: Closing — all 4 compound regularization variants failed badly (best 7.925%, 71.6% worse than 4.619% baseline). Key insight: WD=1e-2 is 100x more aggressive than the baseline WD=1e-4, catastrophically suppressing model capacity. DrivAerML's golden config fragility is further confirmed. Your analysis was clean and thorough. You'll be reassigned to TandemFoil where there's more headroom, frieren..
